pkg/controller/scheduler.go

```diff
@@ -322,32 +325,29 @@ func (c *Controller) advanceCanary(name string, namespace string) {
 
 	// route traffic back to primary if analysis has succeeded
 	if cd.Status.Phase == flaggerv1.CanaryPhasePromoting {
+		//start traffic shift only when the primary is ready
+		if err := canaryController.IsPrimaryReady(cd); err != nil {
+			c.recordEventWarningf(cd, "%v", err)
+			return
+		}
+
 		c.runPromotionTrafficShift(cd, canaryController, meshRouter, provider, canaryWeight, primaryWeight)
 		return
 	}
 
 	// scale canary to zero if promotion has finished
-	if cd.Status.Phase == flaggerv1.CanaryPhaseFinalising {
-		if err := canaryController.ScaleToZero(cd); err != nil {
-			c.recordEventWarningf(cd, "%v", err)
-			return
-		}
+	if (cd.Status.Phase == flaggerv1.CanaryPhaseFinalising ||
+		cd.Status.Phase == flaggerv1.CanaryPhaseWaitingFinalising) &&
```

Comment:
Why are we checking for `CanaryPhaseFinalising`? From what I understand `runFinalizing`, runs the finalizing webhooks, sets the status to `Finalizing`, scales the canary deployment down to zero and then sets the status to `Succeeded`. If `runFinalizing` fails to scale down the canary deployment, then the canary would be in `Finalizing` status, and `runFinalizing()` would be called on the next run, which would subsequently call the finalizing webhooks as well.